_posts/2020-05-26-introduction-to-spring-webflux.md

```diff
@@ -0,0 +1,170 @@
+
+Finally, you'll also need to take into account the somewhat steep learning curve with WebFlux. Its functional, declarative, non-blocking programming model is not easy to start with if you have no prior experience with reactive programming. However, as with many other things, it's a matter of taste and after the learning curve I've grown fond of Webflux's functional style.
+
+## My experiences with Spring WebFlux
+
+I've been using Spring WebFlux in my project for over a year. Why did we end up choosing WebFlux and not traditional MVC?
+- Our application was going to be built from scratch without burden of legacy code
+- Our application mostly consists of HTTP calls to various backend systems, combining and manipulating data to a more user friendly format
+- Data storage is mostly done in other backend systems
+
+Regarding the last point, our data storage needs are not very complex nor are they the main functionality of our application. We chose to isolate our data storage functionality to a dedicated microservice written with Spring MVC and traditional blocking JDBC. Other microservices requiring data storage communicate with this microservice over REST API.  With this way we could write other parts with reactive WebFlux, evaluate the maturity of Spring R2DBC as a side track and maybe later migrate the data storage microservice to WebFlux & R2DBC.
+
+There was indeed a learning curve with WebFlux and its functional programming style. It took me couple of weeks to get comfortable with it and start seeing the benefits. For example, when porting functionality from the old codebase to our new application, it was quite easy to see problematic patterns or clear architectural violations in the old codebase, as trying to implement similar functionality with WebFlux was surprisingly hard. It was clear quite quick that if you need to use old codebase as a reference, you only check what you need to do (e.g. which calls are needed to backend systems) but not how to actually do it.
+
+### Tips for some common pitfalls you might encounter
+
+#### Returning an empty value
+Sometimes you just need to check something in a function, without returning a value. For example, you might need a function to check if some operation can be made. Your function returns a void value in a success case and throws an exception in an error case:
```

Comment:
'Your function returns a void value in case of a success and throws an exception in case of an error'